packages/permission_handler/tizen/src/permission_manager.cc

```diff
@@ -108,16 +110,83 @@ void PermissionManager::RequestPermissions(
   }
 
   on_going_ = true;
-  request_success_callback_ = success_callback;
-  request_error_callback_ = error_callback;
-  result = ppm_request_permissions(permissions_to_request.data(),
-                                   permissions_to_request.size(),
-                                   OnRequestPermissionsResponse, this);
-  if (result != PRIVACY_PRIVILEGE_MANAGER_ERROR_NONE) {
+
+  struct Param {
+    PermissionManager* self{nullptr};
+    bool is_done{false};
+    size_t remaining_request{0};
+  };
+  Param p;
+  p.self = this;
+  p.remaining_request = permissions_to_request.size();
+
+  for (size_t i = 0; i < permissions_to_request.size(); i++) {
+    const char* permission = permissions_to_request[i];
+    p.is_done = false;
+    result = ppm_request_permission(
+        permission,
+        [](ppm_call_cause_e cause, ppm_request_result_e result,
+           const char* privilege, void* data) {
+          Param* param = (Param*)data;
+          PermissionManager* self = param->self;
+
+          if (cause != PRIVACY_PRIVILEGE_MANAGER_CALL_CAUSE_ANSWER) {
+            // abandon a request
+            LOG_ERROR("Privilege[%s] request failed with an error", privilege);
+            param->is_done = true;
+            return;
+          }
+
+          int permission = self->ConvertToPermission(privilege);
+
+          if (self->request_results_.count(permission) == 0) {
+            switch (result) {
+              case PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_ALLOW_FOREVER:
+                self->request_results_[permission] = PERMISSION_STATUS_GRANTED;
+                break;
+              case PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE:
+                self->request_results_[permission] = PERMISSION_STATUS_DENIED;
+                break;
+              case PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_FOREVER:
+                self->request_results_[permission] =
+                    PERMISSION_STATUS_NEVER_ASK_AGAIN;
+                break;
+            }
+          }
+          LOG_DEBUG("permission %d status: %d", permission,
+                    self->request_results_[permission]);
+          auto location =
+              self->request_results_.find(PERMISSION_GROUP_LOCATION);
+          if (location != self->request_results_.end()) {
+            self->request_results_[PERMISSION_GROUP_LOCATION_ALWAYS] =
+                location->second;
+            self->request_results_[PERMISSION_GROUP_LOCATION_WHEN_IN_USE] =
+                location->second;
+          }
+
+          param->remaining_request--;
+          param->is_done = true;
+        },
+        &p);
+
+    if (result != PRIVACY_PRIVILEGE_MANAGER_ERROR_NONE) {
+      LOG_ERROR("Failed to call ppm_request_permission with [%s]", permission);
+      continue;
+    }
+
+    // Wait until ppm_request_permission is done;
+    while (!p.is_done) {
+      ecore_main_loop_iterate();
```

Comment:
As we said in our conversation, I don't think this is strange. I'm not sure if ``timeout`` is necessary here. I think we should listen to other people's opinions.

Reply:
If you have other reasonable ideas, please share it :)

Reply:
It seems only God can use this function according to its documentation. 🤔

https://docs.tizen.org/application/native/api/wearable/latest/group__Ecore__Main__Loop__Group.html#ga0b23ea5d7f31f45e2c2f37d4393ddb2e

Reply:
What a funny document! :laughing: 
I'm not a god, but there's nothing much the user can do when show pop-up for a permission, so I thought it was OK.